run tests with race detector enabled

aojea opened the pull request:
**- What this PR does and why is it needed**

Enable the golang race detector for tests and fixes the race conditions detected.

Fixes: https://github.com/ovn-org/ovn-kubernetes/issues/1368

**- Special notes for reviewers**

**- How to verify it**

Test should pass

**- Description for the changelog**

aojea commented:
/assign @girishmg @danwinship

dcbw commented:
@aojea 

```
2020-07-09T20:59:31.4249527Z github.com/ovn-org/ovn-kubernetes/go-controller/hybrid-overlay/cmd/hybrid-overlay-node
2020-07-09T20:59:35.2507028Z + COVERALLS=1
2020-07-09T20:59:35.2507537Z + make check
2020-07-09T20:59:35.2522202Z hack/test-go.sh 
2020-07-09T20:59:35.5497673Z -covermode must be "atomic", not "set", when -race is enabled
2020-07-09T20:59:35.5504716Z make: *** [check] Error 1
2020-07-09T20:59:35.5508037Z ##[error]Makefile:30: recipe for 
```

aojea commented:
upps, at least I said 😅 

> it´s failing locally 🤔

let me add another job, one with coverage and other with race detection, what do you think?

vjayaramrh commented:
> upps, at least I said 😅
> 
> > it´s failing locally 🤔
> 
> let me add another job, one with coverage and other with race detection, what do you think?

Are you planning to add a CI action for race detection? If you end up doing that, please close the issue https://github.com/ovn-org/ovn-kubernetes/issues/1368 when your PR merges. Thanks

aojea commented:
/retest

aojea commented:
/hold

this is interesting, adding the locks are adding more delays to the operations and some tests that depends on the order of the events are failing now more than before, i.e.

```
• Failure [0.006 seconds]
Informer Event Handler Tests
/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/informer/informer_test.go:54
  adds existing pod and processes a delete event [It]
  /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/informer/informer_test.go:189
```

we´ll continue flaky hunting in this PR

aojea commented:
/retest

aojea commented:
/retest

aojea commented:
/retest

aojea commented:
it hit the same test failure twice
```
OVN NetworkPolicy Low-Level Operations
computes match strings from address sets correctly
```

aojea commented:
/retest

danwinship commented on go-controller/pkg/informer/informer_test.go:
```diff
@@ -113,8 +113,8 @@ var _ = Describe("Informer Event Handler Tests", func() {
-		Consistently(deletes).Should(Equal(int32(0)), "deletes")
-		Eventually(adds, 3).Should(Equal(int32(1)), "adds")
+		Consistently(atomic.LoadInt32(&deletes)).Should(Equal(int32(0)), "deletes")
+		Eventually(atomic.LoadInt32(&adds), 3).Should(Equal(int32(1)), "adds")
```
This code (both original and modified) is not using `Consistently`/`Eventually` correctly; the argument gets evaluated when you first call `Consistently`/`Eventually`, so in both versions, it is only asserting about the initial values of `deletes`/`adds`, not their eventual values. You need to say:

    Consistently(func() int32 { return atomic.LoadInt32(&deletes) }).Should(Equal(int32(0)), "deletes")
    Eventually(func() int32 { return atomic.LoadInt32(&adds) }).Should(Equal(int32(1)), "adds")

although if this is not already flaky, then something must be forcing it to serialize already and we don't need the async versions...

Likewise for some of the other examples modified below, and probably some others nearby as well.

aojea replied:
TIL, https://medium.com/@william.la.martin/ginkgotchas-yeh-also-gomega-13e39185ec96although

> flaky, then something must be forcing it to serialize already and we don't need the async versions...

it seems is already flaky

> Likewise for some of the other examples modified below, and probably some others nearby as well.

heh, seems that this is incorrect in most part of the repo :shrug:

aojea commented on go-controller/hack/test-go.sh:
```diff
@@ -17,16 +17,24 @@ function testrun {
     args="${args}${otherargs}${pkg}"
 
-    go test -mod vendor ${args} ${ginkgoargs}
+    go test -v ${args} ${ginkgoargs} -mod vendor
```
`-mod vendor` fails if used before `-race`

> also, the addition of -v / -ginkgo.v is not related to enabling the race detector is it?

yeah, but your are totally blind without those, there is no way to understand what fails

aojea commented on go-controller/pkg/factory/factory_test.go:
```diff
@@ -378,6 +378,7 @@ var _ = Describe("Watch Factory Operations", func() {
 
 		const nodeName string = "mynode"
 		type opTest struct {
+			mu      sync.Mutex
```
this is the whole stack trace, the race detector complains when the test checks the opTest type variable, because it was modified recently by the handler
https://pastebin.com/NAqaVNtP

aojea commented:
- split in different commits
- fixed ginkgo `Consistently/Eventually` usage

Also, understood why enabling the race detector there are more test failures, it consumes more CPU and memory and the test that are using asynchronous assertions that are not configured correctly are failing because everything is slower

danwinship commented on go-controller/pkg/factory/factory_test.go:
```diff
@@ -558,7 +576,9 @@ var _ = Describe("Watch Factory Operations", func() {
 				nodes = append(nodes, ot.node)
 				nodeWatch.Add(ot.node)
+				ot.mu.Lock()
 				ot.node.Status.Phase = v1.NodeTerminated
```
OK, so the race checker appears to be complaining that we write to this variable here after having read it previously in the `UpdateFunc` above. That is, it is complaining that the _second_ time we do `ot.node.Status.Phase = v1.NodeTerminated` might happen before, after, or during the _first_ time we run `UpdateFunc`.

That is, in the main goroutine we do
```
1. nodeWatch.Add
2. ot.node.Status.Phase = v1.NodeTerminated
3. nodeWatch.Modify
4. nodeWatch.Delete
5. nodeWatch.Add
6. ot.node.Status.Phase = v1.NodeTerminated
7. nodeWatch.Modify
8. nodeWatch.Delete
```

and step 3 guarantees that we eventually call `UpdateFunc`, but doesn't guarantee whether we call it before or after step 6. As it happens, it doesn't matter whether we call it before or after step 6, since step 6 is a no-op. The only actual problem would occur if we called `UpdateFunc` _during_ step 6, and adding a mutex ensures that that can't happen. 

Honestly, it's not clear what the point of this code is anyway; we never test that `Phase` _isn't_ set to `NodeTerminated` anywhere, so we could just move this assignment to before the loop and get the same result...

dcbw replied:
IIRC the point was just to have *something* to update on Node. Not that the value mattered.

danwinship replied:
But it's not testing anything useful. It's just ensuring that `UpdateFunc()` does not get called before the first time we call `nodeWatch.Modify()`, and _nothing else_. It's not like we're testing that updates don't get queued when nothing has changed, or something like that. We could just as easily be using an unrelated global variable here.

aojea commented on go-controller/pkg/factory/factory_test.go:
```diff
@@ -571,16 +591,19 @@ var _ = Describe("Watch Factory Operations", func() {
 		Eventually(c.getDeleted, 2).Should(Equal(10))
 		for _, ot := range testNodes {
+			ot.mu.Lock()
 			Expect(ot.added).Should(Equal(2))
```
the async assert is not done in the 3 lines before the loop?
```
		Eventually(c.getAdded, 2).Should(Equal(10))
		Eventually(c.getUpdated, 2).Should(Equal(10))
		Eventually(c.getDeleted, 2).Should(Equal(10))
```

danwinship replied:
Huh. The Gomega authors apparently don't understand what "asynchronous" means. I assumed that "asynchronous assertions" ran asynchronously, but in fact, `Eventually()` just blocks until it times out or the condition passes. Nice.

danwinship replied:
(Well, ok... it's not "an asynchronous assertion that something happens", it's "a (synchronous) assertion that something happens asynchronously". :man_shrugging: )

dcbw replied:
Eventually() was the way to know when our end state was reached in the tests. We try to do something, and then wait a while and eventually (hah!) that thing would be true, or the test would fail. Which would be the same general scheme as if we just waited for the queues to (eventually!) be drained...

danwinship replied:
except that this way requires us to know for sure how long it will take for the queues to drain in the worst case, and Antonio already had to update the timeouts somewhere (maybe in a different PR?) because they were too short in some cases

aojea replied:
I think that you refer to this PR https://github.com/ovn-org/ovn-kubernetes/pull/1498 , that was the problem, there were no polling so if the queue was't drained before the assert the test failed

dcbw commented:
> It seems like the factory tests (and maybe the informer tests as well) are completely broken, and need to be heavily rethought... we need some factory method to tell us when it has processed all updates, so that we can ensure that we do the end state testing after all callbacks have been run.

@danwinship I'm not sure there's any way to do that. What does "process all updates" actually mean? That every queue is empty? You also don't know when the next event will arrive (though with unit tests can kinda control that).

danwinship commented:
I am only talking about the case of unit tests. Yes, "process all updates" means every queue is empty.
